### PR TITLE
add game logic

### PR DESCRIPTION
**Pull request by LoxiasCam**
### Milestone 3: Game Logic

In this milestone we added working game logic to the game. You can play the game by choosing the cell you want to place your symbol in, for example A1 or C3.

- Added working logic for game flow
- Created classes to handle different aspects of game flow
- User interaction is being handled from main.rb
- Game checks for invalid turns and draw conditions

### CODE REVIEW

File: lib/board.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    return false unless valid_position?(cell_id)
+
+    @grid[cell_id] = player.token
+
+    true
+  end
+
+  def show_board
+    %w[1 2 3].each { |i| p @grid.select { |k| k.include? i }.values }
+  end
```
Comment (BrianSammit) [inline lib/board.rb]: In line 17 you have one `p`. Remember that the print/puts must be only on `main.rb`

File: lib/game.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    # max. number of players is 2
+    return false if @players.length >= 2
+
+    players[name] = Player.new(name, token)
+  end
+
+  def show_board
+    p 'This is the current board:'
+    @board.show_board
+  end
```
Comment (BrianSammit) [inline lib/game.rb]: In line 19 you have one `p`. Remember that the print/puts must be only on `main.rb`

File: bin/main.rb [hunk 3]
```diff
@@ -1,31 +1,37 @@
...
+require_relative '../lib/game'
+require_relative '../lib/player'
 
 p '------------------------'
 p 'Welcome to Tic Tac Toe!'
 p '------------------------'
 p 'Starting new game . . .'
 
-# game = Game.new
+# initialize game
```
Comment (BrianSammit) [inline bin/main.rb]: Remove the inline comment from line 11.

File: lib/game.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+  def initialize
+    @players = {}
+    @board = Board.new
+    @winner = false
+    @turns = 1
+  end
+
+  def add_player(name, token)
+    # max. number of players is 2
```
Comment (BrianSammit) [inline lib/game.rb]: Remove the comment from line 12.

File: lib/game.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  def show_board
+    p 'This is the current board:'
+    @board.show_board
+  end
+
+  def check_for_win
+    @winner = @board.winner?
+  end
+
+  # Check if game has a winner
```
Comment (BrianSammit) [inline lib/game.rb]: Remove the comment from line 27.

File: lib/game.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    false
+  end
+
+  def set_winner
+    @winner = true
+  end
+
+  # Executes a new turn by a player
+  # Returns: boolean
```
Comment (BrianSammit) [inline lib/game.rb:33]: Remove the comments from lines 39 and 40.

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGES REQUIRED :recycle: :red_circle: 
--------------
Hello team!! :wave: Great job working on this milestone :muscle: 
However, there are a few changes to make to complete this milestone :hammer_and_wrench: see them below :point_down: 
--------------

- [x] No puts/print and gets.chomp outside main.rb file. You have two `p` outside the `main.rb` file. See my inline comments :eyes:  to know where are the `p`. kindly move them to `main.rb`
only_main_print

- [x] No inline comments in this milestone. Remove all your inline comments from the code. See my inline comments :eyes: 
----------------
Make these minor changes :hammer_and_wrench: and send it for another review :muscle: 
HAPPY CODING!!
MB5W4D

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Camilo Patino Alzate & Carlos Castillo Velasquez ** 
Nice job with your first submission for this project :muscle:  :muscle:
Your game is already working perfectly :clap: :clap: :clap:  
There are just a few details to improve in your code before you can go ahead:

- [x] Fix rubocop offenses
- [x] Whenever there is a winner, display the winner's name
- [x] Validate player1 and player2 have different names
- [x] Validate player's name is never an empty string
- [x] In Player class, name and token attribute don't need the attr_accessor since they are never written outside the class.
- [x] In Board class, grid attribute doesn't need the attr_reader since it is never read outside the class.
- [x] In Game class, set_winner should be removed if not used, or it should be private if it is never called outside the class.
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Review by bini-i — APPROVED** [review]
# Status: Approved :heavy_check_mark:
### Hello  @LoxiasCam and @ey-charles :wave: 
### Very nice project. Thank you for working on the suggested changes and the project is approved :tada:
#### You can now merge this pull request and continue to the next milestone.

#### If you have any question kindly reach me on slack @Binyam Hailemeskel

### HAPPY CODING :computer: :blush: 

> #### Reviewed by Binyam Hailemeskel (TSE)
> GitHub: [@bini-i](https://github.com/bini-i)
> Twitter: [@binyamshewa](https://twitter.com/binyamshewa)
> LinkedIn: [LinkedIn](https://www.linkedin.com/in/binyam-hailemeskel-728048151/)

**Pull request merged**